src/librustc_trans/attributes.rs

```diff
@@ -134,3 +135,97 @@ pub fn from_fn_attrs(ccx: &CrateContext, attrs: &[ast::Attribute], llfn: ValueRe
 fn cstr(s: &'static str) -> &CStr {
     CStr::from_bytes_with_nul(s.as_bytes()).expect("null-terminated string")
 }
+
+pub fn provide(providers: &mut Providers) {
+    providers.target_features_whitelist = |tcx, cnum| {
+        assert_eq!(cnum, LOCAL_CRATE);
+        Rc::new(llvm_util::target_feature_whitelist(tcx.sess)
+            .iter()
+            .map(|c| c.to_str().unwrap().to_string())
+            .collect())
+    };
+
+    providers.target_features_enabled = |tcx, id| {
+        let whitelist = tcx.target_features_whitelist(LOCAL_CRATE);
+        let mut target_features = Vec::new();
+        for attr in tcx.get_attrs(id).iter() {
+            if !attr.check_name("target_feature") {
+                continue
+            }
+            if let Some(val) = attr.value_str() {
+                for feat in val.as_str().split(",").map(|f| f.trim()) {
+                    if !feat.is_empty() && !feat.contains('\0') {
+                        target_features.push(feat.to_string());
+                    }
+                }
+                let msg = "#[target_feature = \"..\"] is deprecated and will \
+                           eventually be removed, use \
+                           #[target_feature(enable = \"..\")] instead";
+                tcx.sess.span_warn(attr.span, &msg);
+                continue
+            }
+
+            if tcx.fn_sig(id).unsafety() == Unsafety::Normal {
+                let msg = "#[target_feature(..)] can only be applied to \
+                           `unsafe` function";
+                tcx.sess.span_err(attr.span, msg);
```

Comment:
I almost missed this: it's a very bad idea to have these checks here (as opposed to `rustc_typeck`) because they'd only trigger on monomorphizations.

Reply:
Do you have a suggestion for where this might go? This is the same bug about validating `#[inline(never)]` and such I think? (in the sense that this isn't the first of it's kind and it's not clear if there's an obvious way to fix this)

Reply:
`inline` and some aspects of `repr` (whether they are applied to the right kind of item, there are further checks in rustc_typeck) are checked in `rustc::hir::check_attr`. It seems these checks would fit in nicely there, since they doen't seem to need any type information.

Reply:
Thanks for the pointer! Needed a bit of refactoring but it fits there now.